File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    head = ListNode()
    list = head

    while list1 or list2:
      if list1 and (not list2 or list1.val < list2.val):
        list.next = list1
        list = list1
        list1 = list1.next
      elif list2:
        list.next = list2
        list = list2
        list2 = list2.next

    return head.next
        

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
  head := &ListNode{}
  tail := head

  for list1 != nil && list2 != nil {
    if list1.Val < list2.Val {
      tail.Next = list1
      list1 = list1.Next
    } else {
      tail.Next = list2
      list2 = list2.Next
    }
    tail = tail.Next
  }

  tail.Next = list1
  if list2 != nil {
    tail.Next = list2
  }

  return head.Next
}